command/operator_raft_snapshot_inspect.go

```diff
@@ -0,0 +1,669 @@
+	}
+
+	// Generate structs for the formatter with information we read in
+	metaformat := &MetadataInfo{
+		ID:      meta.ID,
+		Size:    meta.Size,
+		Index:   meta.Index,
+		Term:    meta.Term,
+		Version: meta.Version,
+	}
+
+	// Restructures stats given above to be human readable
+	formattedStatsKV := generateKVStats(info)
+
+	in := &OutputFormat{
+		Meta:         metaformat,
+		StatsKV:      formattedStatsKV,
+		TotalCountKV: info.TotalCountKV,
+		TotalSizeKV:  info.TotalSizeKV,
+	}
+
+	out, err := formatter.Format(in)
+	if err != nil {
+		c.UI.Error(err.Error())
+		return 1
+	}
+
+	c.UI.Output(out)
+	return 0
+}
+
+func (c *OperatorRaftSnapshotInspectCommand) kvEnhance(val *pb.StorageEntry, info *SnapshotInfo, read int) {
+	if c.details {
+		if val.Key == "" {
+			return
+		}
+
+		// check for whether a filter is specified. if it is, skip
+		// any keys that don't match.
+		if len(c.filter) > 0 && !strings.HasPrefix(val.Key, c.filter) {
+			return
+		}
+
+		split := strings.Split(string(val.Key), "/")
+
+		// handle the situation where the key is shorter than
+		// the specified depth.
+		actualDepth := c.depth
+		if c.depth == 0 {
+			actualDepth = len(split)
+		} else if c.depth > len(split) {
+			actualDepth = len(split)
+		}
+
+		prefix := strings.Join(split[0:actualDepth], "/")
+		kvs := info.StatsKV[prefix]
+		if kvs.Name == "" {
+			kvs.Name = prefix
+		}
+
+		kvs.Count++
+		kvs.Size += read
+		info.TotalCountKV++
+		info.TotalSizeKV += read
+		info.StatsKV[prefix] = kvs
+	}
+}
+
+func (c *OperatorRaftSnapshotInspectCommand) enhance(file io.Reader) (SnapshotInfo, error) {
+	info := SnapshotInfo{
+		StatsKV:      make(map[string]typeStats),
+		TotalCountKV: 0,
+		TotalSizeKV:  0,
+	}
+
+	handler := func(s *pb.StorageEntry, read int) error {
+		c.kvEnhance(s, &info, read)
+		return nil
+	}
+
+	_, err := ReadSnapshot(file, handler)
+	if err != nil {
+		return info, err
+	}
+
+	return info, nil
+}
+
+// ReadSnapshot decodes each message type and utilizes the handler function to
+// process each message type individually
+func ReadSnapshot(r io.Reader, handler func(s *pb.StorageEntry, read int) error) (*iradix.Tree, error) {
+	reader := protoio.NewDelimitedReader(r, math.MaxInt32)
+
+	errCh := make(chan error, 1)
+	txn := iradix.New().Txn()
+
+	go func() {
+		for {
+			s := new(pb.StorageEntry)
+
+			err := reader.ReadMsg(s)
+			if err != nil {
+				if err == io.EOF {
+					errCh <- nil
+					return
+				}
+				errCh <- err
+				return
+			}
+
+			size := reader.GetLastReadSize()
+
+			handler(s, size)
+
+			var value interface{} = struct{}{}
+			value = s.Value
+
+			txn.Insert([]byte(s.Key), value)
+		}
+	}()
+
+	err := <-errCh
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+
+	return txn.Commit(), nil
+}
+
+// Read a snapshot into a temporary file. Return file and metadata from snapshot.
+// The caller is responsible for removing the file.
+func Read(logger hclog.Logger, in io.Reader) (*os.File, *raft.SnapshotMeta, error) {
```

Comment:
Why do we need the temporary file?  The vault server code does something like when doing a restore, but in our case we're not receiving the snapshot as part of an HTTP request, we're not streaming it, our input is already a file on disk.

Reply:
We write the state.bin from the snapshot to the temporary file. That comment is misleading. I'll update

Reply:
Removed the temporary file 👍